File: NeuroEvolution/flappyBird/agent.js
class Agent {

    constructor(initBrain, initEmpty) {

        // If you change the gravity and lift then the velocity limits also need changing. This is max & minLim
        this.yPos = height / 2;
        this.xPos = 50;
        this.gravity = 0.6;
        this.lift = -15;
        this.maxLim = 6;
        this.minLim = -15;
        this.velocity = 0;
        this.radius = 25;

        this.timeSamplesExperianced = 0;
        this.totalDistanceFromGapOverTime = 0;

        this.score = 0;
        this.fitness = 0;
        this.avgDistFromGap = 0;

        if (initBrain.CLASS_TAG !== undefined || initBrain === true) {
            if (initBrain.CLASS_TAG === "nn.js") {

                this.brain = NeuralNetwork.deserialize(initBrain);
            }

        } else {

            let msLayerUnits = [6, 4, 2];
            let msActFunctions = ["RELU", "softmax"];

            this.brain = new NeuralNetwork(msLayerUnits, msActFunctions);

            if (initEmpty === false) {
                this.brain.initLayers("he_normal");
            } else {
                this.brain.initLayers(); // init an empty array of elements
            }
        }

    }

    show() {
        fill(0, 100);
        stroke(0.5);
        ellipse(this.xPos, this.yPos, this.radius, this.radius);
    }

    think(clostestBlock) {

        // If the screen is too small not enough blocks will spawn meaning
        // that clostestBlock will stay null. So just increase screen size this error will go away
        // this should be a consideration when using pixel input.

        let inputs = [];
        inputs[0] = clostestBlock.xPos / width;
        inputs[1] = clostestBlock.topStart / height;
        inputs[2] = clostestBlock.bottomStart / height;
        inputs[3] = (clostestBlock.xPos - this.xPos) / width; // This the distance from agent to the next block
        inputs[4] = this.yPos / height;
        inputs[5] = this.minMaxNormalise(this.velocity); // Need to Normalise this some how?? Use an activation function like sigmoid
        // it can squash the numbers but if the velocity is too big it will ceiling???? 

        let prediction = this.brain.feedFoward(inputs);

        // console.log(prediction.data[0] + "  " + prediction.data[1]);
        if (prediction.data[0] > prediction.data[1]) {
            this.actionUp();
        }

        // determine which pipe is closest
        // 

    }

    actionUp() {
        this.velocity += this.lift;
    }

    update(clostestBlock) {
        //  The velcoity cumlatively increases with respect to the amount of gravity
        // Update the yPosition with the velocity. y = the top is 0 and the bottom of the screen is the height 
        // of the window. 

        this.velocity += this.gravity;
        this.velocity *= 0.9;
        this.yPos += this.velocity;

        if (this.velocity > this.maxLim) {
            this.velocity = this.maxLim;
        }

        if (this.velocity < this.minLim) {
            this.velocity = this.minLim;
        }
        // Stop the agent from falling past the screen limits
        if (this.yPos > height) {
            this.yPos = height;
            this.velocity = 0;

        } else if (this.yPos < 0) {
            this.yPos = 0;
            this.velocity = 0;

        }

        // penelise agents for their distance on the y from the center of the gap of the obsticles
        let gap = clostestBlock.bottomStart - clostestBlock.topStart;
        let gapMid = clostestBlock.topStart + Math.round((gap / 2));
        let agentDistanceFromGap = Math.floor(Math.abs(this.yPos - gapMid));

        this.totalDistanceFromGapOverTime += agentDistanceFromGap;
        this.timeSamplesExperianced++;

        this.score++;
    }

    minMaxNormalise(x) {
        return (x - this.minLim) / (this.maxLim - this.minLim);
    }

    computeFitness() {

        // penalise agent based on avg distance from gap

        let impactFactor = 0.5; // adjusts the percentage of penalisation applied
        this.avgDistFromGap = Math.floor(this.totalDistanceFromGapOverTime / this.timeSamplesExperianced);
        this.score -= Math.floor(impactFactor * this.avgDistFromGap);
        if (this.score < 0) {
            this.score = 0;
        }
    }


}



if (typeof module !== 'undefined') {
    module.exports = Agent;
}

File: python_models/src_STM_model/src_pycharm/agent.py
import pygame
import numpy as np
from nn import NeuralNetwork


class Agent(pygame.sprite.Sprite):

    def __init__(self, initBrain, initEmpty, screen_width, screen_height, name='agent'):
        # Call the parent's constructor
        super().__init__()

        self.gravity  = 0.6
        self.lift     = -2
        self.maxLim   = 6
        self.minLim   = -15
        self.velocity = 0
        self.radius   = 2
        self.color = (0, 0, 0, 50)

        self.name = name
        self.image = pygame.Surface([self.radius, self.radius], pygame.SRCALPHA)
        self.image.fill(self.color)
        self.rect = self.image.get_rect()
        self.rect.x = 10
        self.rect.y = screen_height/2

        self.timeSamplesExperianced       = 0
        self.totalDistanceFromGapOverTime = 0

        self.fitness        = 0
        self.avgDistFromGap = 0

        # NEED TO ADD LOAD IN BRAIN FUNCTIONS BELOW:
        #
        #
        #
        #

        msLayeruUnits = [9, 4, 2]
        msActFunctions = ["relu", "softmax"]

        self.brain = NeuralNetwork(layer_units=msLayeruUnits, activation_func_list=msActFunctions)

        if initEmpty == False:
            self.brain.init_layers(init_type="he_normal")

        else:
            self.brain.init_layers(init_type="zeros")



    def show(self):
        pass


    def think(self, closestBlock, screen_width, screen_height):

        inputs = []
        # input about on coming block object
        inputs.append(closestBlock.xPos / screen_width)
        inputs.append(closestBlock.top_block.rect.bottom / screen_height)
        inputs.append(closestBlock.bottom_block.rect.top / screen_height)
        inputs.append((closestBlock.xPos - self.rect.right) / screen_width)

        # input about agents current position
        inputs.append(self.rect.bottom / screen_height)
        inputs.append(self.rect.top / screen_height)
        inputs.append(self.rect.right / screen_height)
        inputs.append(self.rect.left / screen_height)
        inputs.append(self.minMaxNormalise(self.velocity))

        inputs = np.array([inputs])

        prediction = self.brain.feed_foward(inputs=inputs)

        if prediction[0] > prediction[1]:
            self.actionUp()


    def actionUp(self):
        self.velocity += self.lift


    def update(self, closestBlock, screen_height):

        self.velocity += self.gravity
        self.velocity *= 0.9
        self.rect.y += self.velocity

        if self.velocity > self.maxLim:
            self.velocity = self.maxLim

        if self.velocity < self.minLim:
            self.velocity = self.minLim

        if self.rect.bottom > screen_height:
            self.rect.bottom = screen_height
            self.velocity = 0

        elif self.rect.top < 5:
            self.rect.top = 0
            self.velocity = 0

        # penalise agents for their distance on the y from the center of the gap of the blocks
        gap = closestBlock.bottom_block.rect.top - closestBlock.top_block.rect.bottom
        gapMid = closestBlock.top_block.rect.bottom + np.round((gap / 2))
        agentDistanceFromGap = np.floor(np.abs(self.rect.midright[1] - gapMid))

        self.totalDistanceFromGapOverTime = self.totalDistanceFromGapOverTime + agentDistanceFromGap
        self.timeSamplesExperianced = self.timeSamplesExperianced + 1

        self.fitness = self.fitness + 1


    def minMaxNormalise(self, x):
        return (x - self.minLim) / (self.maxLim - self.minLim)


    def computeFitness(self):
        # penalise agent based on average distance from gap

        impactFactor = 0.5 # scales the percentage of penalisation applied
        self.avgDistFromGap = np.floor(self.totalDistanceFromGapOverTime / self.timeSamplesExperianced)
        self.fitness = self.fitness - np.floor(impactFactor * self.avgDistFromGap)
        if self.fitness < 0:
            self.fitness = 0


if __name__ == "__main__":
    pass




